### PR TITLE
Form alert validation

### PR DESCRIPTION
**Pull request by ChitvanRamani22**
Closes: #107 
@Namanl2001 if any of the required field (as per the schema) is empty an alert will pop up and the form wont get submitted.

### CODE REVIEW

**Comment by Namanl2001** [comment]
Hey @ChitvanRamani22 your changes look nice but we are trying to achieve something like [this](https://user-images.githubusercontent.com/41958505/112602110-482b2800-8e39-11eb-8d9b-47068e1c6d44.png). As it looks nicer and user friendly. WDYT? In alert we are not aware of which input field we are missing.

**Comment by ChitvanRamani22** [comment]
@Namanl2001 this is what i have already done in my previous pr and there you commemted that in issue 107 we need a alert of any of the required field as per the model is missing.
So thats what i did.

**Comment by ChitvanRamani22** [comment]
I can do one thing that is  print the messeges of requirement of that particular input below that input field.
Will that be fine @Namanl2001

**Comment by ChitvanRamani22** [comment]
@Namanl2001 please have a look at this commit and let me know if this correct.

**Comment by Namanl2001** [comment]
We can consider this but it does not have good UI/UX and looks lil clumsy at first. WDYT?
Have you tried to implement this? https://user-images.githubusercontent.com/41958505/112602110-482b2800-8e39-11eb-8d9b-47068e1c6d44.png

**Comment by ChitvanRamani22** [comment]
@Namanl2001 the image shows the same thing that at starting till the fields are empty the messege will be displayed else it will disappear.

**Comment by Namanl2001** [comment]
> @Namanl2001 the image shows the same thing that at starting till the fields are empty the messege will be displayed else it will disappear.

Nope, it's not like that. The warnings are only shown when a user clicks the submit form button, else there is no err/warning.

**Comment by Namanl2001** [comment]
Did you find any solution for this?

**Comment by ChitvanRamani22** [comment]
@Namanl2001 not yet.

**Comment by ChitvanRamani22** [comment]
@Namanl2001 please review now.

**Comment by Namanl2001** [comment]
Looks more good now 👍 Few things

- can we have toast msg instead of windows.alert() ?
- make `name` compulsory
- Styling  of the form is being altered

**Comment by ChitvanRamani22** [comment]
@Namanl2001 name is compulsory.
But as the name input field is not empty the astrik sign is not visible

**Comment by ChitvanRamani22** [comment]
@Namanl2001 toast messege is not working as should.

**Comment by ChitvanRamani22** [comment]
@Namanl2001 i have added toast messege instead of alert.

**Comment by Namanl2001** [comment]
> @Namanl2001 i have added toast messege instead of alert.

will this err be shown on the form header? Please attach screenshot

File: src/components/form.jsx [hunk 1]
```diff
@@ -57,48 +76,89 @@ class form extends Component {
       this.setState({ errormessage2: err });
     }
 
-    this.setState({
-      [e.target.id]: e.target.value,
-    });
+    switch (nam) {
+      case 'title':
+        errors.title = val == '' ? '*' : '';
```
Comment (Namanl2001) [inline src/components/form.jsx:81]:
```suggestion
        errors.title = val === '' ? '*' : '';
```

File: src/components/form.jsx [hunk 2]
```diff
@@ -57,48 +76,89 @@ class form extends Component {
...
+        break;
+      case 'userName':
+        errors.userName = val.length < 1 ? '*' : '';
+        break;
+      case 'subject':
+        errors.subject = val == '' ? '*' : '';
```
Comment (Namanl2001) [inline src/components/form.jsx:87]: duplicate

File: src/components/form.jsx [hunk 3]
```diff
@@ -57,48 +76,89 @@ class form extends Component {
...
+        break;
+      case 'tutor':
+        errors.tutor = val == '' ? '*' : '';
```
Comment (Namanl2001) [inline src/components/form.jsx:90]: duplicate

File: src/components/update.jsx [hunk 4]
```diff
@@ -47,57 +69,98 @@ class update extends Component {
...
         );
         document.getElementById('phone').style.border = '1px solid red';
       } else {
         document.getElementById('phone').style.borderColor = '';
       }
       this.setState({ errormessage2: err });
     }
+    switch (nam) {
+      case 'title':
+        errors.title = val == '' ? '*' : '';
```
Comment (Namanl2001) [inline src/components/update.jsx:84]: duplicate

File: src/components/update.jsx [hunk 5]
```diff
@@ -47,57 +69,98 @@ class update extends Component {
...
+        break;
+      case 'userName':
+        errors.userName = val.length < 1 ? '*' : '';
+        break;
+      case 'subject':
+        errors.subject = val == '' ? '*' : '';
```
Comment (Namanl2001) [inline src/components/update.jsx:90]: duplicate

File: src/components/update.jsx [hunk 6]
```diff
@@ -47,57 +69,98 @@ class update extends Component {
...
+        break;
+      case 'tutor':
+        errors.tutor = val == '' ? '*' : '';
```
Comment (Namanl2001) [inline src/components/update.jsx:93]: duplicate

**Review by Namanl2001 — CHANGES_REQUESTED** [review]
make [this](https://github.com/Namanl2001/MERN-Gurujii-dev/blob/40a022c0c87d19118b600d13211512d94c57c3b0/src/components/style.css#L469) `margin-right`.

Also make [this](https://github.com/Namanl2001/MERN-Gurujii-dev/blob/40a022c0c87d19118b600d13211512d94c57c3b0/src/components/update.jsx#L237) `col='sm-2'` -> `col='sm-3'` for all the four classes in update.jsx

The above nits are for the styling part else LGTM. But the main thing of concern here is after these changes I'm not able to update the user profile, looks like a bug. Please look into it.

Thanks

**Comment by ChitvanRamani22** [comment]
> make [this](https://github.com/Namanl2001/MERN-Gurujii-dev/blob/40a022c0c87d19118b600d13211512d94c57c3b0/src/components/style.css#L469) `margin-right`.
> 
> Also make [this](https://github.com/Namanl2001/MERN-Gurujii-dev/blob/40a022c0c87d19118b600d13211512d94c57c3b0/src/components/update.jsx#L237) `col='sm-2'` -> `col='sm-3'` for all the four classes in update.jsx
> 
> The above nits are for the styling part else LGTM. But the main thing of concern here is after these changes I'm not able to update the user profile, looks like a bug. Please look into it.
> 
> Thanks

@Namanl2001 can you please tell in which line number i want to change it to margin-right.

**Comment by Namanl2001** [comment]
Click on attached hyperlinks in the above comment.

**Comment by ChitvanRamani22** [comment]
@Namanl2001 now everything is working fine!😀

**Comment by Namanl2001** [comment]
Please resolve the conflicts then I will merge this tomorrow. Thanks

**Comment by ChitvanRamani22** [comment]
@Namanl2001 can you please merge this now.

**Comment by Namanl2001** [comment]
check is not passing yet. Please look into it

**Comment by ChitvanRamani22** [comment]
@Namanl2001 i dont know why that chck is not passing i havent made any change in user.controller file.

**Comment by Namanl2001** [comment]
PHAL @ [this](https://github.com/Namanl2001/MERN-Gurujii-dev/pull/18#issuecomment-824134125)

**Review by Namanl2001 — COMMENTED** [review]
Let's do this. And check

**Review by Namanl2001 — CHANGES_REQUESTED** [review]
Small nit.
Also please comment [this line](https://github.com/Namanl2001/MERN-Gurujii-dev/blob/2521487e367afbda9421fa91cdf127973931f53c/backend/controllers/user.controller.js#L99) for now.

also after doing these changes please heck once if all the functions ae working properly ene to end and as expected.
Thanks

File: src/components/form.jsx [hunk 7]
```diff
@@ -58,49 +77,91 @@ class form extends Component {
...
-            )
-          ) {
-            window.location.reload();
+      axios
+        .post('/users/add', data)
+        .then(response => {
+          if (response.status === 200) {
+            axios.get(`/users/sendMail/${this.props.emailid}/1`);
+            if (
+              alert(
```
Comment (Namanl2001) [inline src/components/form.jsx:151]:
```suggestion
             ! alert(
```

**Review by Namanl2001 — CHANGES_REQUESTED** [review]
Small nit.
Also please comment [this line](https://github.com/Namanl2001/MERN-Gurujii-dev/blob/2521487e367afbda9421fa91cdf127973931f53c/backend/controllers/user.controller.js#L99) for now.

also after doing these changes please heck once if all the functions ae working properly ene to end and as expected.
Thanks

**Comment by ChitvanRamani22** [comment]
@Namanl2001 i have done the changes.
And everything is working fine now.

**Review by Namanl2001 — APPROVED** [review]
Thanks @ChitvanRamani22 for your patience and great work. 🚀

**Pull request merged**